clippy_lints/src/duplicate_match_guards.rs

```diff
@@ -0,0 +1,171 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::{
+    HasSession, SpanRangeExt, indent_of, reindent_multiline, snippet_with_applicability, trim_span,
+};
+use clippy_utils::{eq_expr_value, span_contains_comment};
+use rustc_errors::Applicability;
+use rustc_hir::{Arm, ExprKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::BytePos;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for the same condition being checked in a match guard and in the match body
+    ///
+    /// ### Why is this bad?
+    /// This is usually just a typo or a copy and paste error.
+    ///
+    /// ### Known problems
+    /// False negatives: if the condition is an impure function, it could've been called twice on
+    /// purpose for its side effects
+    ///
+    /// ### Example
+    /// ```no_run
+    /// # let n = 0;
+    /// # let a = 3;
+    /// # let b = 4;
+    /// match n {
+    ///     0 if a > b => {
+    ///         if a > b {
+    ///             return;
+    ///         }
+    ///     }
+    ///     _ => {}
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// # let n = 0;
+    /// # let a = 3;
+    /// # let b = 4;
+    /// match n {
+    ///     0 if a > b => {
+    ///         return;
+    ///     }
+    ///     _ => {}
+    /// }
+    /// ```
+    #[clippy::version = "1.89.0"]
+    pub DUPLICATE_MATCH_GUARDS,
```

Comment:
Couldn't this lint in principle apply to regular if conditions too? E.g. this currently has no warning
```rs
let x = 1;
if x > 1 {
  if x > 1 {
    // always executes
  } else {
    // this block of code is simply unreachable
    todo!()
  }
} else { todo!() }
```
I'm not saying that this PR should implement that, but it makes me wonder if we should make the lint name a bit more general, like `duplicate_nested_conditions` or something so it could be extended in the future without needing a name change?